From 11bce22f520b3d6a621669f40c2ae41b47f9cca7 Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:39:44 -0400
Subject: [virt] kvm/x86: IN instruction emulation should ignore REP-prefix

Message-id: <1406040016-3289-190-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86268
O-Subject: [RHEL7 PATCH v2 189/221] KVM: x86: IN instruction emulation should ignore REP-prefix
Bugzilla: 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Nadav Amit <namit@cs.technion.ac.il>

The IN instruction is not be affected by REP-prefix as INS is.  Therefore, the
emulation should ignore the REP prefix as well.  The current emulator
implementation tries to perform writeback when IN instruction with REP-prefix
is emulated. This causes it to perform wrong memory write or spurious #GP
exception to be injected to the guest.

Signed-off-by: Nadav Amit <namit@cs.technion.ac.il>
Signed-off-by: Marcelo Tosatti <mtosatti@redhat.com>
(cherry picked from commit e6e39f0438bc4b0da9334ca42337775c7a00db21)
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/emulate.c b/arch/x86/kvm/emulate.c
index f3834bb..e8a5840 100644
--- a/arch/x86/kvm/emulate.c
+++ b/arch/x86/kvm/emulate.c
@@ -1324,7 +1324,8 @@ static int pio_in_emulated(struct x86_emulate_ctxt *ctxt,
   rc->end = n * size;
  }
 
- if (ctxt->rep_prefix && !(ctxt->eflags & EFLG_DF)) {
+ if (ctxt->rep_prefix && (ctxt->d & String) &&
+     !(ctxt->eflags & EFLG_DF)) {
   ctxt->dst.data = rc->data + rc->pos;
   ctxt->dst.type = OP_MEM_STR;
   ctxt->dst.count = (rc->end - rc->pos) / size;
-- 
1.7.1